### PR TITLE
Feature/task defaults

### PR DESCRIPTION
**Pull request by jlariono**
- Added support for setting task parameter default values using a `TaskDefaults` global property in a Whiskey configuration file.
- Added `Add-WhiskeyTaskDefault` function for adding new task default parameter values to an existing task context object.
- Misc test fixes and improvements

### CODE REVIEW

**Comment by jlariono** [comment]
I'll submit changes after I can rebase my work on top of the changes in PR #132, since that PR affects some of the files in this PR.

**Comment by jlariono** [comment]
PR ready for review now. Task default functionality moved into a `TaskDefaults` task.

File: Whiskey/Functions/Add-WhiskeyTaskDefault.ps1 [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    Use-CallerPreference -Cmdlet $PSCmdlet -SessionState $ExecutionContext.SessionState
+
+    if (-not ($Context | Get-Member -Name 'TaskDefaults'))
+    {
+        throw 'The given ''Context'' object does not contain a ''TaskDefaults'' property. Create a proper Whiskey context object using the ''New-WhiskeyContext'' function.'
+    }
+
+    if ($Task -notin (Get-WhiskeyTask | Select-Object -ExpandProperty 'Name'))
+    {
+        throw 'The Task ''{0}'' is not a valid Whiskey task.' -f $Task
```
Comment (splatteredbits) [inline Whiskey/Functions/Add-WhiskeyTaskDefault.ps1]: Change error message to `Task ''{0}'' does not exist.`

File: Whiskey/Functions/New-WhiskeyContext.ps1 [hunk 2]
```diff
@@ -8,19 +8,28 @@ function New-WhiskeyContext
...
-    * `OutputDirectory`: the path to a directory where build output, reports, etc. should be saved. This directory is created for you.
-    * `Version`: a `SemVersion.SemanticVersion` object representing the semantic version to use when building the application. This object has two extended properties: `Version`, a `Version` object that represents the semantic version with all pre-release and build metadata stripped off; and `ReleaseVersion` a `SemVersion.SemanticVersion` object with all build metadata stripped off.
-    * `ReleaseVersion`: the semantic version with all build metadata stripped away, i.e. the version and pre-release only.
-    * `Configuration`: the parsed YAML as a hashtable.
-    * `DownloadRoot`: the path to a directory where tools can be downloaded when needed. 
     * `ByBuildServer`: a flag indicating if the build is being run by a build server.
     * `ByDeveloper`: a flag indicating if the build is being run by a developer.
-    * `ApplicatoinName`: the name of the application being built.
+    * `BuildMetadata`: metadata information of the current build on a build server.
```
Comment (splatteredbits) [inline Whiskey/Functions/New-WhiskeyContext.ps1]: Some of these properties are private and shouldn't be used outside of Whiskey. Remove:

* Configuration
* Credentials
* DownloadRoot
* Task Defaults
* Variables

File: Test/Invoke-WhiskeyTask.Tests.ps1 [hunk 3]
```diff
@@ -584,36 +617,30 @@ function WhenRunningTask
...
-    $script:context = New-WhiskeyTestContext @byItDepends
-    $context.PipelineName = 'Build';
-    $context.TaskName = $null;
-    $context.TaskIndex = 1;
-    $context.TaskDefaults = $taskDefaults;
-
     if( $InRunMode )
     {
-        $context.RunMode = $InRunMode;
+        $script:context.RunMode = $InRunMode
```
Comment (splatteredbits) [inline Test/Invoke-WhiskeyTask.Tests.ps1]: You only need the `$script:` scope when setting the value of the variable, not properties on the variable.

File: Test/Invoke-WhiskeyTask.Tests.ps1 [hunk 4]
```diff
@@ -3,6 +3,8 @@ Set-StrictMode -Version 'Latest'
```
Comment (splatteredbits) [inline Test/Invoke-WhiskeyTask.Tests.ps1]: I don't understand why you made any changes to this test. You only changed one line in Invoke-WhiskeyTask that shouldn't have had any affect on tests. Can you give me some context?

File: Test/WhiskeyTest.psm1 [hunk 5]
```diff
@@ -294,12 +290,9 @@ function New-WhiskeyTestContext
         }
     }
 
-    $context = New-WhiskeyContextObject
+    $context = New-WhiskeyContext -Environment 'Verification' -ConfigurationPath $ConfigurationPath
```
Comment (splatteredbits) [inline Test/WhiskeyTest.psm1]: Why this change? It requires any test that calls it to setup an entire build environment, which is expensive and makes tests have to know everything about Whiskey instead of just what they care about.

Comment (jlariono) [reply Test/WhiskeyTest.psm1]: Oh this change was originally made when I initially made `TaskDefaults` a global config property that was set in `New-WhiskeyContext`. Now that `TaskDefaults` is a task, this is no longer necessary. I'll revert this.

File: Test/Invoke-WhiskeyTask.Tests.ps1 (hunk 4 above)
Comment (jlariono) [reply Test/Invoke-WhiskeyTask.Tests.ps1]: When I was making `TaskDefaults` a global config I added some new tests to this file at which point I also did some refactoring to make the tests cleaner. I ended up not needing the tests that I added, but left the refactoring. Should I revert these changes?

Comment (splatteredbits) [reply Test/Invoke-WhiskeyTask.Tests.ps1]: Yes, please.

**Comment by jlariono** [comment]
@splatteredbits Requested changes completed. Ready for re-review.

**Comment by jlariono** [comment]
@splatteredbits Realized a Whiskey release has occurred since I created this branch. I rebased onto develop, incremented the version number, and removed the old release notes. I'll need another approval.

**Comment by splatteredbits** [comment]
@jlariono OK. Let me know when you're ready.

**Comment by jlariono** [comment]
@splatteredbits It's ready for review now. I rebased onto develop and the only file changed was `Whiskey.psd1`. I incremented the `Whiskey.psd1` version number when resolving the merge conflicts during the rebase.

**Comment by splatteredbits** [comment]
@jlariono I didn't see any new commits so couldn't easily see what changed.
